src/debug_wait.c

```diff
@@ -44,23 +45,64 @@ ts_debug_waitpoint_init(DebugWait *waitpoint, const char *tagname)
  * other sessions that try to grab the same lock but will block if an
  * exclusive lock is already taken, and then release the lock immediately
  * after.
+ *
+ * This function can decide to block while taking the shared lock or it can
+ * have a retry loop to take the share lock. This retry loop option is useful
+ * in cases where this function gets called from deep down inside a transaction
+ * where interrupts are not being served currently.
  */
 void
-ts_debug_waitpoint_wait(DebugWait *waitpoint)
+ts_debug_waitpoint_wait(DebugWait *waitpoint, bool blocking)
 {
 	LockAcquireResult lock_acquire_result pg_attribute_unused();
 	bool lock_release_result pg_attribute_unused();
 
-	ereport(DEBUG1, (errmsg("waiting on waitpoint '%s'", waitpoint->tagname)));
-
-	/* Take the lock. This should always succeed, anything else is a bug. */
-	lock_acquire_result = LockAcquire(&waitpoint->tag, ShareLock, true, false);
+	ereport(DEBUG3, (errmsg("waiting on waitpoint '%s'", waitpoint->tagname)));
+
+	if (blocking)
+		lock_acquire_result = LockAcquire(&waitpoint->tag, ShareLock, true, false);
+	else
+	{
+		/*
+		 * Trying to wait indefinitely here could lead to hangs. The current
+		 * behavior is to retry for retry_count and return with a warning
+		 * if that's crossed.
+		 *
+		 * If required, in future, we could take an additional option to decide
+		 * if the caller wants to retry indefinitely or return with a warning.
+		 * But the current behavior based on the "blocking" argument is ok for
+		 * now.
+		 */
+		unsigned int retry_count = 1000;
+
+		/* try to acquire the lock without waiting. */
+		lock_acquire_result = LockAcquire(&waitpoint->tag, ShareLock, true, true);
```

Comment:
since this statement is repeated at the end of the loop, maybe we can just have one statement within the loop instead, and then break out or if it on the right conditions?

Reply:
Ok, using a `do..while` loop now.